**Descriptive Analysis**

Imports

In [1]:
import pandas as pd

In [2]:
import os, sys
sys.path.append(os.path.abspath(os.path.join('../scripts')))
from Data_Selector import *
from Data_Cleaner import *
from Data_Visualizer import *
from Data_Outlier_Handler import OutlierHandler

**Data Reading**

In [3]:
# Reading a file with a custom delimiter
df = pd.read_csv('../Data/MachineLearningRating_v3.txt', delimiter='|')

C:\Users\USER\AppData\Local\Temp\ipykernel_20400\2279570384.py:2: DtypeWarning: Columns (32,37) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../Data/MachineLearningRating_v3.txt', delimiter='|')


**Data Understanding**

In [4]:
df.head()

UnderwrittenCoverID  PolicyID     TransactionMonth  IsVATRegistered  \
0               145249     12827  2015-03-01 00:00:00             True   
1               145249     12827  2015-05-01 00:00:00             True   
2               145249     12827  2015-07-01 00:00:00             True   
3               145255     12827  2015-05-01 00:00:00             True   
4               145255     12827  2015-07-01 00:00:00             True   

  Citizenship          LegalType Title Language                 Bank  \
0              Close Corporation    Mr  English  First National Bank   
1              Close Corporation    Mr  English  First National Bank   
2              Close Corporation    Mr  English  First National Bank   
3              Close Corporation    Mr  English  First National Bank   
4              Close Corporation    Mr  English  First National Bank   

       AccountType  ...                    ExcessSelected CoverCategory  \
0  Current account  ...             Mobility - Windscreen    Windscreen   
1  Current account  ...             Mobility - Windscreen    Windscreen   
2  Current account  ...             Mobility - Windscreen    Windscreen   
3  Current account  ...  Mobility - Metered Taxis - R2000    Own damage   
4  Current account  ...  Mobility - Metered Taxis - R2000    Own damage   

    CoverType            CoverGroup              Section  \
0  Windscreen  Comprehensive - Taxi  Motor Comprehensive   
1  Windscreen  Comprehensive - Taxi  Motor Comprehensive   
2  Windscreen  Comprehensive - Taxi  Motor Comprehensive   
3  Own Damage  Comprehensive - Taxi  Motor Comprehensive   
4  Own Damage  Comprehensive - Taxi  Motor Comprehensive   

                           Product StatutoryClass StatutoryRiskType  \
0  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
1  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
2  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
3  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
4  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   

   TotalPremium TotalClaims  
0     21.929825         0.0  
1     21.929825         0.0  
2      0.000000         0.0  
3    512.848070         0.0  
4      0.000000         0.0  

[5 rows x 52 columns]

In [5]:
df.tail()

UnderwrittenCoverID  PolicyID     TransactionMonth  IsVATRegistered  \
1000093                31520       389  2015-04-01 00:00:00            False   
1000094                31520       389  2015-06-01 00:00:00            False   
1000095                31520       389  2015-08-01 00:00:00            False   
1000096                31519       389  2014-07-01 00:00:00            False   
1000097                31519       389  2015-02-01 00:00:00            False   

        Citizenship   LegalType Title Language       Bank      AccountType  \
1000093          ZW  Individual    Mr  English  ABSA Bank  Savings account   
1000094          ZW  Individual    Mr  English  ABSA Bank  Savings account   
1000095          ZW  Individual    Mr  English  ABSA Bank  Savings account   
1000096          ZW  Individual    Mr  English  ABSA Bank  Savings account   
1000097          ZW  Individual    Mr  English  ABSA Bank  Savings account   

         ... ExcessSelected        CoverCategory            CoverType  \
1000093  ...      No excess          Third Party          Third Party   
1000094  ...      No excess          Third Party          Third Party   
1000095  ...      No excess          Third Party          Third Party   
1000096  ...      No excess  Passenger Liability  Passenger Liability   
1000097  ...      No excess  Passenger Liability  Passenger Liability   

                   CoverGroup              Section  \
1000093  Comprehensive - Taxi  Motor Comprehensive   
1000094  Comprehensive - Taxi  Motor Comprehensive   
1000095  Comprehensive - Taxi  Motor Comprehensive   
1000096  Comprehensive - Taxi  Motor Comprehensive   
1000097  Comprehensive - Taxi  Motor Comprehensive   

                                    Product StatutoryClass StatutoryRiskType  \
1000093  Mobility Commercial Cover: Monthly     Commercial     IFRS Constant   
1000094  Mobility Commercial Cover: Monthly     Commercial     IFRS Constant   
1000095  Mobility Commercial Cover: Monthly     Commercial     IFRS Constant   
1000096  Mobility Commercial Cover: Monthly     Commercial     IFRS Constant   
1000097  Mobility Commercial Cover: Monthly     Commercial     IFRS Constant   

         TotalPremium TotalClaims  
1000093    347.235175         0.0  
1000094    347.235175         0.0  
1000095    347.235175         0.0  
1000096      2.315000         0.0  
1000097      2.315000         0.0  

[5 rows x 52 columns]

In [6]:
df.size

52005096

In [7]:
df.shape

(1000098, 52)

In [8]:
df.dtypes

UnderwrittenCoverID           int64
PolicyID                      int64
TransactionMonth             object
IsVATRegistered                bool
Citizenship                  object
LegalType                    object
Title                        object
Language                     object
Bank                         object
AccountType                  object
MaritalStatus                object
Gender                       object
Country                      object
Province                     object
PostalCode                    int64
MainCrestaZone               object
SubCrestaZone                object
ItemType                     object
mmcode                      float64
VehicleType                  object
RegistrationYear              int64
make                         object
Model                        object
Cylinders                   float64
cubiccapacity               float64
kilowatts                   float64
bodytype                     object
NumberOfDoors               

**Summary**

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000098 entries, 0 to 1000097
Data columns (total 52 columns):
 #   Column                    Non-Null Count    Dtype  
---  ------                    --------------    -----  
 0   UnderwrittenCoverID       1000098 non-null  int64  
 1   PolicyID                  1000098 non-null  int64  
 2   TransactionMonth          1000098 non-null  object 
 3   IsVATRegistered           1000098 non-null  bool   
 4   Citizenship               1000098 non-null  object 
 5   LegalType                 1000098 non-null  object 
 6   Title                     1000098 non-null  object 
 7   Language                  1000098 non-null  object 
 8   Bank                      854137 non-null   object 
 9   AccountType               959866 non-null   object 
 10  MaritalStatus             991839 non-null   object 
 11  Gender                    990562 non-null   object 
 12  Country                   1000098 non-null  object 
 13  Province                  1

**Statistical Info**

In [10]:
df.describe()

UnderwrittenCoverID      PolicyID    PostalCode        mmcode  \
count         1.000098e+06  1.000098e+06  1.000098e+06  9.995460e+05   
mean          1.048175e+05  7.956682e+03  3.020601e+03  5.487770e+07   
std           6.329371e+04  5.290039e+03  2.649854e+03  1.360381e+07   
min           1.000000e+00  1.400000e+01  1.000000e+00  4.041200e+06   
25%           5.514300e+04  4.500000e+03  8.270000e+02  6.005692e+07   
50%           9.408300e+04  7.071000e+03  2.000000e+03  6.005842e+07   
75%           1.391900e+05  1.107700e+04  4.180000e+03  6.005842e+07   
max           3.011750e+05  2.324600e+04  9.870000e+03  6.506535e+07   

       RegistrationYear      Cylinders  cubiccapacity      kilowatts  \
count      1.000098e+06  999546.000000  999546.000000  999546.000000   
mean       2.010225e+03       4.046642    2466.743258      97.207919   
std        3.261391e+00       0.294020     442.800640      19.393256   
min        1.987000e+03       0.000000       0.000000       0.000000   
25%        2.008000e+03       4.000000    2237.000000      75.000000   
50%        2.011000e+03       4.000000    2694.000000     111.000000   
75%        2.013000e+03       4.000000    2694.000000     111.000000   
max        2.015000e+03      10.000000   12880.000000     309.000000   

       NumberOfDoors  CustomValueEstimate  NumberOfVehiclesInFleet  \
count  999546.000000         2.204560e+05                      0.0   
mean        4.019250         2.255311e+05                      NaN   
std         0.468314         5.645157e+05                      NaN   
min         0.000000         2.000000e+04                      NaN   
25%         4.000000         1.350000e+05                      NaN   
50%         4.000000         2.200000e+05                      NaN   
75%         4.000000         2.800000e+05                      NaN   
max         6.000000         2.655000e+07                      NaN   

         SumInsured  CalculatedPremiumPerTerm  TotalPremium   TotalClaims  
count  1.000098e+06              1.000098e+06  1.000098e+06  1.000098e+06  
mean   6.041727e+05              1.178757e+02  6.190550e+01  6.486119e+01  
std    1.508332e+06              3.997017e+02  2.302845e+02  2.384075e+03  
min    1.000000e-02              0.000000e+00 -7.825768e+02 -1.200241e+04  
25%    5.000000e+03              3.224800e+00  0.000000e+00  0.000000e+00  
50%    7.500000e+03              8.436900e+00  2.178333e+00  0.000000e+00  
75%    2.500000e+05              9.000000e+01  2.192982e+01  0.000000e+00  
max    1.263620e+07              7.442217e+04  6.528260e+04  3.930921e+05

In [11]:
#Checking number of unique rows in each feature

df.nunique().sort_values()

NumberOfVehiclesInFleet          0
Language                         1
ItemType                         1
Country                          1
StatutoryClass                   1
StatutoryRiskType                1
CrossBorder                      1
IsVATRegistered                  2
WrittenOff                       2
NewVehicle                       2
Rebuilt                          2
Converted                        2
AlarmImmobiliser                 2
TrackingDevice                   2
TermFrequency                    2
MaritalStatus                    3
Gender                           3
AccountType                      3
Product                          4
Citizenship                      4
VehicleType                      5
Section                          5
Title                            5
LegalType                        6
NumberOfDoors                    6
Cylinders                        7
Province                         9
Bank                            11
ExcessSelected      

**Missing Value**

In [24]:
df.isna().sum()

UnderwrittenCoverID               0
PolicyID                          0
TransactionMonth                  0
IsVATRegistered                   0
Citizenship                       0
LegalType                         0
Title                             0
Language                          0
Bank                         145961
AccountType                   40232
MaritalStatus                  8259
Gender                         9536
Country                           0
Province                          0
PostalCode                        0
MainCrestaZone                    0
SubCrestaZone                     0
ItemType                          0
mmcode                          552
VehicleType                     552
RegistrationYear                  0
make                            552
Model                           552
Cylinders                       552
cubiccapacity                   552
kilowatts                       552
bodytype                        552
NumberOfDoors               

In [12]:
percent_missing_values(df)

The dataset contains 9.74 % missing values.


In [13]:
missing_df = missing_values_table(df)

Your selected dataframe has 52 columns.
There are 22 columns that have missing values.


In [14]:
missing_df

Missing Values  % of Total Values    Dtype
NumberOfVehiclesInFleet         1000098             100.00  float64
CrossBorder                      999400              99.93   object
CustomValueEstimate              779642              77.96  float64
WrittenOff                       641901              64.18   object
Converted                        641901              64.18   object
Rebuilt                          641901              64.18   object
NewVehicle                       153295              15.33   object
Bank                             145961              14.59   object
AccountType                       40232               4.02   object
Gender                             9536               0.95   object
MaritalStatus                      8259               0.83   object
mmcode                              552               0.06  float64
VehicleType                         552               0.06   object
make                                552               0.06   object
VehicleIntroDate                    552               0.06   object
NumberOfDoors                       552               0.06  float64
bodytype                            552               0.06   object
kilowatts                           552               0.06  float64
cubiccapacity                       552               0.06  float64
Cylinders                           552               0.06  float64
Model                               552               0.06   object
CapitalOutstanding                    2               0.00   object

As a convention, it is good to remove columns with more than 30% of missing values.

In [18]:
columns_to_be_removed = missing_df[missing_df['% of Total Values'] >= 30.00].index.tolist()
columns_to_be_removed

['NumberOfVehiclesInFleet',
 'CrossBorder',
 'CustomValueEstimate',
 'WrittenOff',
 'Converted',
 'Rebuilt']

I'm going to retain columns like CrossBorder, CustomValueEstimate, WrittenOff, and Rebuilt because they could provide insights into risk profiles. However, NumberOfVehiclesInFleet and Converted might not directly influence risk for individual claims and could be excluded.

In [19]:
columns_to_be_removed = [col for col in columns_to_be_removed if col not in ['CrossBorder',
    'WrittenOff', 'CustomValueEstimate', 'Rebuilt']]
columns_to_be_removed

['NumberOfVehiclesInFleet', 'Converted']

In [20]:
cleaned_df = df.drop(columns_to_be_removed, axis=1)
cleaned_df.shape

(1000098, 50)

Now let's check the status of missing values in the modified dataframe.

In [21]:
percent_missing_values(cleaned_df)

The dataset contains 6.85 % missing values.


In [23]:
missing_values_table(cleaned_df)

Your selected dataframe has 50 columns.
There are 20 columns that have missing values.


Missing Values  % of Total Values    Dtype
CrossBorder                  999400              99.93   object
CustomValueEstimate          779642              77.96  float64
WrittenOff                   641901              64.18   object
Rebuilt                      641901              64.18   object
NewVehicle                   153295              15.33   object
Bank                         145961              14.59   object
AccountType                   40232               4.02   object
Gender                         9536               0.95   object
MaritalStatus                  8259               0.83   object
mmcode                          552               0.06  float64
make                            552               0.06   object
VehicleType                     552               0.06   object
bodytype                        552               0.06   object
kilowatts                       552               0.06  float64
cubiccapacity                   552               0.06  float64
Cylinders                       552               0.06  float64
Model                           552               0.06   object
NumberOfDoors                   552               0.06  float64
VehicleIntroDate                552               0.06   object
CapitalOutstanding                2               0.00   object

In [46]:
missing_count = cleaned_df['CrossBorder'].isna().sum()
print(f"Missing values in CrossBorder: {missing_count}")

Missing values in CrossBorder: 999400


Since CrossBorder column is almost all null we will remove it.

In [47]:
drop_columns(cleaned_df, 'CrossBorder')

11 columns were dropped.


Since "Rebuilt" , "NewVehicle" , "Gender", "MaritalStatus"and "WrittenOff"  are categorical columns, it is better to impute them with the value "unknown" so that we don't bias the data.

In [48]:
fix_missing_value(cleaned_df, 'Rebuilt', 'unknown')
fix_missing_value(cleaned_df, 'NewVehicle', 'unknown')
fix_missing_value(cleaned_df, 'WrittenOff', 'unknown')
fix_missing_value(cleaned_df, 'Gender', 'unknown')
fix_missing_value(cleaned_df, 'Bank', 'unknown')
fix_missing_value(cleaned_df, 'MaritalStatus', 'unknown')

641901 missing values in the column Rebuilt have been replaced by unknown.
153295 missing values in the column NewVehicle have been replaced by unknown.
641901 missing values in the column WrittenOff have been replaced by unknown.
9536 missing values in the column Gender have been replaced by unknown.
145961 missing values in the column Bank have been replaced by unknown.
8259 missing values in the column MaritalStatus have been replaced by unknown.


0          Not specified
1          Not specified
2          Not specified
3          Not specified
4          Not specified
               ...      
1000093           Single
1000094           Single
1000095           Single
1000096           Single
1000097           Single
Name: MaritalStatus, Length: 1000098, dtype: object

Since the missing percentage of CustomValueEstimate is very high we will impute them with the backward fill method. In this case, using a single value like mean or median is not advisable as it may heavily bais our data in an unwanted way by making the majority of values equal to a single value.

In [51]:
fix_missing_bfill(cleaned_df, 'CustomValueEstimate')

c:\Users\USER\Desktop\KAIM\AlphaCare-Insurance-Analysis\scripts\Data_Cleaner.py:93: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df[col] = df[col].fillna(method='bfill')


779642 missing values in the column CustomValueEstimate have been replaced using the backward fill method.


0          119300.0
1          119300.0
2          119300.0
3          119300.0
4          119300.0
             ...   
1000093         NaN
1000094         NaN
1000095         NaN
1000096         NaN
1000097         NaN
Name: CustomValueEstimate, Length: 1000098, dtype: float64

In [52]:
percent_missing_values(cleaned_df)

The dataset contains 0.09 % missing values.


In [53]:
percent_missing_rows(cleaned_df)

4.08 % of the rows in the dataset contain atleast one missing value.


Since only 4.08% of the rows contain atleast one missing value and the total number of rows is around 150000, dropping these rows won't have a noticable negative impact.

In [54]:
drop_rows_with_missing_values(cleaned_df)

40840 rows containg missing values were dropped.


In [55]:
percent_missing_values(cleaned_df)

The dataset contains 0.0 % missing values.


Now there is no missing value.